File: internal/files/settings_generations.go
package files

import (
	"fmt"
	"os"
	"path/filepath"
)

// settingsGenerations is the FIXED allowlist of settings.json generations the
// diff/restore panel may read — never an arbitrary path.
var settingsGenerations = []string{
	"settings.json",
	"settings.json.bak",
	"settings.json.pre-ponytail",
}

func isSettingsGeneration(name string) bool {
	for _, g := range settingsGenerations {
		if g == name {
			return true
		}
	}
	return false
}

// ListSettingsGenerations returns the allowlisted generations that exist on disk.
func ListSettingsGenerations() ([]string, error) {
	cd, err := claudeDir()
	if err != nil {
		return nil, err
	}
	out := []string{}
	for _, g := range settingsGenerations {
		if info, err := os.Lstat(filepath.Join(cd, g)); err == nil && !info.IsDir() {
			out = append(out, g)
		}
	}
	return out, nil
}

// ReadSettingsGeneration returns an allowlisted generation's raw JSON text (or
// "" if absent). Refuses any name outside the allowlist.
func ReadSettingsGeneration(name string) (string, error) {
	if !isSettingsGeneration(name) {
		return "", fmt.Errorf("files: %q is not a settings generation", name)
	}
	cd, err := claudeDir()
	if err != nil {
		return "", err
	}
	data, err := os.ReadFile(filepath.Join(cd, name))
	if os.IsNotExist(err) {
		return "", nil
	}
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// RestoreSettingsGeneration overwrites settings.json with the chosen generation
// through ReplaceSettingsJSON (current → .bak as-is, new parse-validated, atomic).
func RestoreSettingsGeneration(name string) error {
	if !isSettingsGeneration(name) {
		return fmt.Errorf("files: %q is not a settings generation", name)
	}
	if name == "settings.json" {
		return fmt.Errorf("files: cannot restore settings.json onto itself")
	}
	cd, err := claudeDir()
	if err != nil {
		return err
	}
	data, err := os.ReadFile(filepath.Join(cd, name))
	if err != nil {
		return fmt.Errorf("files: read generation %q: %w", name, err)
	}
	return ReplaceSettingsJSON(data)
}

File: src-tauri/src/files/settings_generations.rs
//! Ports `internal/files/settings_generations.go` — allowlisted settings.json
//! "generations" the diff/restore panel may read. Every entry point gates on the
//! FIXED `SETTINGS_GENERATIONS` allowlist; an arbitrary path is never accepted.

use std::fs;
use std::io;

use super::settings_write::replace_settings_json;
use crate::config::root::claude_dir;

/// The FIXED allowlist of settings.json generations the diff/restore panel may
/// read — never an arbitrary path. Mirrors `settingsGenerations`.
const SETTINGS_GENERATIONS: [&str; 3] = [
    "settings.json",
    "settings.json.bak",
    "settings.json.pre-ponytail",
];

fn is_settings_generation(name: &str) -> bool {
    SETTINGS_GENERATIONS.contains(&name)
}

/// Returns the allowlisted generations that exist on disk. Mirrors
/// `ListSettingsGenerations`.
pub fn list_settings_generations() -> Result<Vec<String>, String> {
    let cd = claude_dir()?;
    let mut out = Vec::new();
    for g in SETTINGS_GENERATIONS {
        if let Ok(info) = fs::symlink_metadata(cd.join(g)) {
            if !info.is_dir() {
                out.push(g.to_string());
            }
        }
    }
    Ok(out)
}

/// Returns an allowlisted generation's raw JSON text (or "" if absent). Refuses
/// any name outside the allowlist. Mirrors `ReadSettingsGeneration`.
pub fn read_settings_generation(name: &str) -> Result<String, String> {
    if !is_settings_generation(name) {
        return Err(format!("files: {name:?} is not a settings generation"));
    }
    let cd = claude_dir()?;
    match fs::read(cd.join(name)) {
        Ok(data) => Ok(String::from_utf8_lossy(&data).into_owned()),
        Err(e) if e.kind() == io::ErrorKind::NotFound => Ok(String::new()),
        Err(e) => Err(e.to_string()),
    }
}

/// Overwrites settings.json with the chosen generation through
/// `replace_settings_json` (current → .bak as-is, new parse-validated, atomic).
/// Mirrors `RestoreSettingsGeneration`.
pub fn restore_settings_generation(name: &str) -> Result<(), String> {
    if !is_settings_generation(name) {
        return Err(format!("files: {name:?} is not a settings generation"));
    }
    if name == "settings.json" {
        return Err("files: cannot restore settings.json onto itself".to_string());
    }
    let cd = claude_dir()?;
    let data = fs::read(cd.join(name)).map_err(|e| format!("files: read generation {name:?}: {e}"))?;
    replace_settings_json(&data)
}

#[cfg(test)]
#[path = "settings_generations_tests.rs"]
mod settings_generations_tests;
